Be more prettier

sirreal opened the pull request:
- Use size-optimized build of prettier, see https://github.com/Automattic/wp-calypso/pull/29987 for details. Same behavior. Faster, smaller installs (props @jsnajdr 👏)
- Add compiled files to `.eslintignore`. We're not interested in linting or prettifying compiled files. Prettier is set to use this for its ignore.
- Add `*.min.js` to `.eslintignore`. We don't want to linting or prettify minified JS.
- Run prettier on _all the JS_ (with a few exceptions).

 ✨ Think about it.

## Questions

Do we need to drop jshint in favor of Calypso eslint? Calypso eslint has evolved with prettier so that they complement each other. We drop any lints that are purely formatting (prettier automates this) and only show lints that require developer intervention.

Do we need to audit the differences in the configs (current jshint vs. Calypso eslint)? To the best of my knowledge, there is some JavaScript in Jetpack that is served, as-is, without transpilation. That's very different from Calypso and may have different linting requirements.

One of the prettier options that we're using in Calypso is `es5` trailing comma, like this:

```js
// none (align with jshint rule: Extra comma. (it breaks older versions of IE))
[ 1, 2 ]; { a: 'a' }; console.log( 1 );
// es5 (current option, what Calypso uses)
[ 1, 2, 3, ]; { a: 'a', }; console.log( 1 );
// all (for completeness, we don't want this at this time)
[ 1, 2, 3, ]; { a: 'a', }; console.log( 1, );
```

https://github.com/Automattic/jetpack/blob/0dc222f5382588ffe0d7c92c4f2618e452db7edd/.prettierrc#L5

## Changelog

- yarn (install the updated prettier - should have identical functionality)
- Run prettier on all of the source JavaScript:
  ```
  yarn run reformat-files
  ```

Closes #11474

sirreal committed:
Run prettier on more files

jeherve commented on package.json:
```diff
@@ -160,7 +160,7 @@
     "nock": "10.0.6",
     "node-wp-i18n": "1.2.3",
-    "prettier": "github:Automattic/wp-prettier#wp-prettier-1.16.4",
+    "prettier": "https://github.com/Automattic/wp-prettier/releases/download/wp-1.16.4/wp-prettier-1.16.4.tgz",
```
What's the reasoning behind that change?

sirreal replied:
It's an optimized bundle of our prettier flavor that provides the bin without any excess. Details: https://github.com/Automattic/wp-calypso/pull/29987

jeherve commented on package.json:
```diff
@@ -54,7 +54,7 @@
     "test-modules": "NODE_ENV=test NODE_PATH=tests:_inc/client:node_modules/@automattic/dops-components/client tests/runner.js modules",
     "test-dangerci-and-adminpage": "yarn danger ci && yarn test-adminpage",
-    "reformat-files": "./node_modules/.bin/prettier --ignore-path .eslintignore --write \"_inc/client/**/*.{js,jsx}\""
+    "reformat-files": "./node_modules/.bin/prettier --ignore-path .eslintignore --write \"_inc/client/**/*.{js,jsx}\" \"modules/**/*.{js,jsx}\""
```
Should we also include js files in the `_inc` directory? We have a few JavaScript files in `_inc/lib` for example.

sirreal replied:
Let's do it all!

jeherve reviewed (commented):
Well this should be fun to sync back to WordPress.com :)

Is this ready for review?

kraftbj commented:
We should see about forcing this in our pre-commit hook for whatever files we prettify: https://github.com/Automattic/jetpack/blob/master/bin/pre-commit-hook.js#L51

kraftbj commented:
Testing failure is due to multiple jshint errors on legacy files.

simison commented:
>We should see about forcing this in our pre-commit hook for whatever files we prettify

Here's an example from Calypso: https://github.com/Automattic/wp-calypso/blob/4e38750947bc346e6438744b755c0ced585530cb/bin/pre-commit-hook.js#L51-L63

simison commented:
>Testing failure is due to multiple jshint errors on legacy files.

@brbrr did you start replacing jshint with eslint as mentioned in Slack?

simison commented:
>We should see about forcing this in our pre-commit hook

There might be some overlap with https://github.com/Automattic/jetpack/pull/11515

sirreal commented:
> Well this should be fun to sync back to WordPress.com :)

😅 Have faith! 😉 

> Is this ready for review?

This was a quick one just to gauge interest —or resistance!

I'll update the description with some real details, run on more JS (all of `_inc`) and label for review.

sirreal commented:
Rebased and ran prettier

sirreal commented:
> One of the prettier options that we're using in Calypso is es5 trailing comma, like this:

I think we can safely drop the jshint rule and use `es5` trailing commas.

According to MDN, trailing commas are supported:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas#Browser_compatibility

2ality says <=ie8 are problematic. I don't think that's in our support matrix 😁 

http://2ality.com/2013/07/trailing-commas.html

sirreal commented:
I've looked at jshint docs and our config and I think we're sorely in need of an update. That's a topic for another PR, but I'd like draw some conclusion from conversation around this PR in the meantime 🙇

sirreal commented:
Rebased and reprettierfied ✨

lezama reviewed (dismissed):
@sirreal you can't be more prettier

enejb commented on modules/custom-css/custom-css/js/codemirror.min.js:
```diff
@@ -4,8 +4,7128 @@
  * Includes CSS & LESS modes.
  * v3.19.1
  */
-window.CodeMirror=function(){"use strict";function x(a,c){if(!(this instanceof x))return new x(a,c);this.options=c=c||{};for(var d in _c)!c.hasOwnProperty(d)&&_c.hasOwnProperty(d)&&(c[d]=_c[d]);J(c);var e="string"==typeof c.value?0:c.value.first,f=this.display=y(a,e);f.wrapper.CodeMirror=this,G(this),c.autofocus&&!p&&Nb(this),this.state={keyMaps:[],overlays:[],modeGen:0,overwrite:!1,focused:!1,su ...
```
@sirreal This file should probably not be prettified. 
I wonder if we should be just using the code mirror file that gets shipped with core instead?

enejb commented on modules/shortcodes/js/jquery.cycle.min.js:
```diff
@@ -6,4 +6,1211 @@
  * Dual licensed under the MIT and GPL licenses.
  * http://jquery.malsup.com/license.html
  */
-!function(e,t){"use strict";function n(t){e.fn.cycle.debug&&i(t)}function i(){window.console&&console.log&&console.log("[cycle] "+Array.prototype.join.call(arguments," "))}function c(t,n,i){var c=e(t).data("cycle.opts");if(c){var s=!!t.cyclePause;s&&c.paused?c.paused(t,c,n,i):!s&&c.resumed&&c.resumed(t,c,n,i)}}function s(n,s,o){function l(t,n,c){if(!t&&n===!0){var s=e(c).data("cycle.opts");if(!s) ...
```
@sirreal This seems to be a jquery plugin http://jquery.malsup.com/cycle/ but it doesn't seems to be maintained. 

Also not it seems like we are prettifying the minified version here.

sirreal replied:
This is used for the slideshow:

https://github.com/Automattic/jetpack/blob/abf3988da482778e9d90ddb35b9ee2368a9d905d/modules/shortcodes/slideshow.php#L275

Yep, would be good to leave this as-is. I'll roll it back add this to `.eslintignore` 👍

enejb commented on modules/shortcodes/js/recipes-printthis.js:
```diff
@@ -35,255 +35,260 @@
  *
  * Notes:
  *  - the loadCSS will load additional css (with or without @media print) into the iframe, adjusting layout
- * 
```
@sirreal  
This seems to be a jquery plugin as well. 
That could also be updated to a newer version see 
https://github.com/jasonday/printThis/blob/master/printThis.js

sirreal replied:
With the source duplicated here, I don't see much reason not to run prettier on this.

Updating the plugin would be material for another PR.

sirreal commented:
Rebased

enejb commented:
Can we also add `modules/custom-css/custom-css/js/codemirror.min.js`  to the .eslintignore rules for now.

sirreal commented:
I've rebased and added `*.min.js` to the `.eslintignore`.

sirreal commented on modules/carousel/jetpack-carousel.js:
```diff
@@ -1552,7 +1552,7 @@ jQuery( document ).ready( function( $ ) {
 					// registered infiniscroll event kicks in, so we don't ever display comments for the wrong image by mistake.
 					var current = $( '.jp-carousel div.selected' );
-					if ( current && current.data && current.data( 'attachment-id' ) != args.attachment_id ) {
+					if ( current && current.data && current.data( 'attachment-id' ) !== args.attachment_id ) {
```
I was concerned about this change but I've tested and confirmed that we're getting numbers on both sides 👍

enejb replied:
Thanks for confirming and checking that!

sirreal replied:
This was ignored before, but the ignore line comment was dropped down by prettier.

I think this is the riskiest linter-initiated change out of the bunch and I thank after rebasing on the other fixes I'd prefer to put the ignore comment back on this line.

sirreal commented:
Can we proceed in spite of the wpcom test failing?

@brbrr do we have some e2e tests that we can run against this?

enejb commented:
@dereksmart Any idea what the best process is to get this merged on the .com side? 
This PR would help us a lot.

scruffian commented:
@kraftbj what kind of review do you want? Just a sanity check that these changes won't change the functionality?

kraftbj commented:
@scruffian Yeah and if anything is questionable at all to verify it is non-functional (at least one strict comparison added was already checked above already, etc).

scruffian commented:
There's a lot to check there! I found a few issues when looking at the Phabricator diff, but I'm not sure if that's just a problem with the way it's displayed...

kraftbj commented:
@scruffian With the phab diff, I'm less concerned. We'll likely need to scrap it and do some file-by-file work. Fusion couldn't apply the patch cleanly, so the breakage is likely due to some things out of sync already messing things up.

If you could be another set of eyes that the PR is 🚢 , we can get it merged, then fight the sync battle moving forward so the PR doesn't get moldy/stale.

sirreal commented:
We might break this PR down into several, it won't take much work but will clarify things greatly.

There are lint fixes here that were prompted by the prettier changes. Those are the riskiest changes and can be isolated.

The eslintignote and prettier dependency changes can also be isolated.

Running prettier can be daunting, but it is very safe. The danger comes from when the large prettier diff obscures hides relevant changes with sheer volume. When isolated into a high volume low risk change, review is easier.

Shall I proceed with that plan?

roccotripaldi commented:
> Shall I proceed with that plan?

Seems like a solid plan!

scruffian commented on modules/custom-css/custom-css/js/css-editor.js:
```diff
@@ -31,53 +31,60 @@
 		$( '#css-mode-select' ).slideUp();
-		$( '#css-mode-display' ).text( $( 'input[name=add_to_existing_display]:checked' ).val() == 'true' ? 'Add-on' : 'Replacement' ); // jshint ignore:line
+		$( '#css-mode-display' ).text(
+			$( 'input[name=add_to_existing_display]:checked' ).val() === 'true' ? 'Add-on' : 'Replacement'
```
Another example of `==` being replaced with `===`

sirreal replied:
This is not prettier, this was a manual change I did to fix the lint.

Broken out into #11576, I'll remove from this PR

scruffian commented:
This looks good to me except the one comment above.

I also think that these files look like libs, so maybe we don't want to change them:

_inc/postmessage.js
_inc/spin.js
modules/custom-css/custom-css/js/core-customizer-css-preview.js
modules/custom-css/custom-css/js/core-customizer-css.core-4.9.js
modules/custom-css/custom-css/js/core-customizer-css.js

Certainly the bottom of _inc/postmessage.js is a copy of json2...

sirreal commented on modules/carousel/jetpack-carousel.js:
```diff
@@ -1138,7 +1138,7 @@ jQuery( document ).ready( function( $ ) {
 			imageLinkParser.href = args.large_file;
 
-			var isPhotonUrl = imageLinkParser.hostname.match( /^i[\d]{1}.wp.com$/i ) != null;
+			var isPhotonUrl = imageLinkParser.hostname.match( /^i[\d]{1}.wp.com$/i ) !== null;
```
https://github.com/Automattic/jetpack/pull/11577

sirreal commented:
I've extracted several smaller parts of this that should be safe to land. I'll rebase and we should wind up with a very large but very safe change here the will only run prettier on more files.

- [ ] https://github.com/Automattic/jetpack/pull/11579
- [ ] https://github.com/Automattic/jetpack/pull/11578
- [ ] https://github.com/Automattic/jetpack/pull/11577
- [ ] https://github.com/Automattic/jetpack/pull/11576

sirreal commented:
Dropped the extracted changes and rebased.

brbrr commented:
> @brbrr did you start replacing jshint with eslint as mentioned in Slack?

@simison Sorry for the delayed response. I was AFK last week. No, I haven't started anything yet.

brbrr commented:
> Can we proceed in spite of the wpcom test failing?
> 
> @brbrr do we have some e2e tests that we can run against this?

Most of our e2e tests are Calypso related, so I'd say no - we don't have e2e tests on which we can rely on here.

sirreal committed:
Run prettier

sirreal commented:
Rebased and conflict free! Let's get pretty!

kraftbj commented:
There's an error from Travis wrt `yarn lint`. Can you give it a look? https://travis-ci.org/Automattic/jetpack/jobs/508925642

With 7.2 code freeze on Monday EOD, I'd like to merge this as early as possible next week _after_ the 7.2 branch is spun off so it has the most time possible to soak in master.

sirreal committed:
Restore disable lint comment to correct line

sirreal commented:
The lint should be fixed, it's a disable-line comment that was shifted when prettier ran.

👍 Sounds like a good plan to land when the 7.2 branch is created.

sirreal committed:
Prettier json!

sirreal committed:
Add some eslint ignores for prettier

sirreal committed:
Remove rule override

Conflicts with prettier formatting of async () => {} function

simison commented:
Let's update `.editorconfig` here: 

https://github.com/Automattic/jetpack/blob/5c63eae84cd232c8031c28676db0db0e2362d010/.editorconfig#L16-L18

simison commented:
Thoughts on updating the pre-commit hook to run on everything here?

https://github.com/Automattic/jetpack/blob/5c63eae84cd232c8031c28676db0db0e2362d010/bin/pre-commit-hook.js#L44-L56

sirreal committed:
Prettier json

sirreal committed:
Reformat files

sirreal commented:
Rebased, addressed feedback:
- Now we prettier json on precommit
- Run prettier on _all the js/json/jsx_ except what's in .eslintignore

Ran prettier again, conflict free.

sirreal commented:
Rebased. With 7.2 branched, this should be ready to land anytime.

ockham commented:
Rebased with :heart: and care.

ockham committed:
Some moar prettier

sirreal commented:
Rebased to fix conflicts

ockham commented:
Need to rebase again, I'll do that. We should land this soon so we're not trapped in rebase limbo forever.

simison commented on .editorconfig:
```diff
@@ -13,7 +13,7 @@ insert_final_newline = true
 indent_style = tab
 
-[{.jshintrc,*.json,*.yml}]
+[{.jshintrc,*.yml}]
```
I think we can remove `.jshintrc` as that file is already using tabs.

simison replied:
did in https://github.com/Automattic/jetpack/pull/11507/commits/f54fd2bf873ac89ea72961297157365285f78b9f

simison commented on modules/custom-post-types/js/many-items.js:
```diff
@@ -1,27 +1,27 @@
 /* jshint onevar: false, smarttabs: true */
```
Noting that I don't see these options in https://jshint.com/docs/options/

Otherwise, they seemed like something that would've conflicted with Prettier.

ockham replied:
Those options[ were removed](https://github.com/jshint/jshint/releases/tag/2.5.0) from `jshint` 2.5.0 four years ago! :laughing: 

JP is on `jshint` 2.9.7 now, probably fine to remove already.

sirreal replied:
```suggestion
```

simison replied:
Doesn't need to be in this PR, seems tedious task as there might be valid options mixed in between.

Seems like easier to remove them all at once together with jshint.

simison reviewed (dismissed):
It's impossible to test _everything_ this file touches but I've poked around, skimmed the changes and all looks good.

Also ran different build and lint scripts everything runs as expected.

It'll have a month to be in `master` before 7.3 ships so it'll get some more testing that way.

simison committed:
Remove jshintrc file in editorconfig

simison committed:
Add jquery.spin.js to ignores

simison committed:
Revert jquery.spin.js

kraftbj committed:
Merge branch 'master' into update/more-prettier

kraftbj reviewed (approved):
Rebased again, need to run prettier on package.json, but avoiding it trying to avoid more rebasing until Travis lands.
